base/weakkeydict.jl

```diff
@@ -2,62 +2,74 @@
 
 type WeakKeyDict{K,V} <: Associative{K,V}
     ht::Dict{Any,V}
-    deleter::Function
+    lock::Threads.RecursiveSpinLock
+    finalizer::Function
 
-    WeakKeyDict() = new(Dict{Any,V}(), identity)
+    function WeakKeyDict()
+        t = new(Dict{Any,V}(), Threads.RecursiveSpinLock(), identity)
+        t.finalizer = function(k)
+            # when a weak key is finalized, remove from dictionary if it is still there
+            islocked(t) && return finalizer(k, t.finalizer)
+            delete!(t, k)
+        end
+        return t
+    end
 end
 WeakKeyDict() = WeakKeyDict{Any,Any}()
 
-function weak_key_delete!(t::Dict, k)
-    # when a weak key is finalized, remove from dictionary if it is still there
-    wk = getkey(t, k, secret_table_token)
-    if !is(wk,secret_table_token) && is(wk.value, k)
-        delete!(t, k)
-    end
-end
+islocked(wkh::WeakKeyDict) = islocked(wkh.lock)
+lock(f, wkh::WeakKeyDict) = lock(f, wkh.lock)
+trylock(f, wkh::WeakKeyDict) = trylock(f, wkh.lock)
 
 function setindex!{K}(wkh::WeakKeyDict{K}, v, key)
-    t = wkh.ht
     k = convert(K, key)
-    if is(wkh.deleter, identity)
-        wkh.deleter = x->weak_key_delete!(t, x)
+    finalizer(k, wkh.finalizer)
```

Comment:
so this may or may not add the finalizer to the input depending ob whether the convert aliased?